godot-cell/src/guards.rs

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::ops::{Deref, DerefMut};
+use std::ptr::NonNull;
+use std::sync::Mutex;
+
+use crate::CellState;
+
+// ----------------------------------------------------------------------------------------------------------------------------------------------
+
+/// Wraps a shared borrowed value of type `T`.
+///
+/// No other mutable borrows to the same value will be created while this guard exists.
+#[derive(Debug)]
+pub struct RefGuard<'a, T> {
+    state: &'a Mutex<CellState<T>>,
+    value: NonNull<T>,
+}
+
+impl<'a, T> RefGuard<'a, T> {
+    /// Create a new `GdRef` guard which can be immutably dereferenced.
+    ///
+    /// # Safety
+    ///
+    /// It must be safe to call [`as_ref()`](NonNull::as_ref) on `value` for as long as the guard is not
+    /// dropped.
+    ///
+    /// In particular you must ensure that:
+    ///
+    /// - The value behind the `value` pointer must be accessible for as long as the guard is not dropped.
+    /// - No new mutable references to the same value can be created.
+    /// - If there exist any other mutable references, then `value` must be derived from those references.
+    /// - Any existing mutable references must stop accessing this value while this guard exists.
+    pub(crate) unsafe fn new(state: &'a Mutex<CellState<T>>, value: NonNull<T>) -> Self {
+        Self { state, value }
+    }
+}
+
+impl<'a, T> Deref for RefGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: It is safe to call `as_ref()` on value because of the safety invariants of `new`.
+        unsafe { self.value.as_ref() }
+    }
+}
+
+impl<'a, T> Drop for RefGuard<'a, T> {
+    fn drop(&mut self) {
+        self.state
+            .lock()
+            .unwrap()
+            .borrow_state
+            .decrement_shared()
+            .unwrap();
+    }
+}
+
+// ----------------------------------------------------------------------------------------------------------------------------------------------
+
+/// Wraps a mutably borrowed value of type `T`.
+///
+/// This prevents all other borrows of `value`, unless the `&mut` reference handed out from this guard is set
+/// as non-aliasing by a call to [`GdCell::set_non_aliasing()`](crate::GdCell::set_non_aliasing).
+#[derive(Debug)]
+pub struct MutGuard<'a, T> {
```

Comment:
I think exclusive doesn't portray the intent much better. Since if i were to set something as exclusive i would imagine that it makes that reference exclusive. i.e it would prevent any other references from being created.

I think non-aliasing maybe has a similar issue here tbh. 

Maybe a name like "reserved" or "captured" or something would be better? something that communicates what we actually do with the reference, rather than talking about the state that the reference is currently in?

Reply:
Good point. There are currently these semantics:

> This prevents all other borrows of `value`, unless the `&mut` reference handed out from this guard is set as "TERM" by a call to `GdCell::set_TERM`

Which means: _if_ TERM is set, then other borrows are possible. However "reserved" or "captured" expresses the opposite: it conveys a state of "locked in", i.e. not being able to be borrowed.

Maybe `set_reborrowable`, `set_shareable`, `set_open`, ...
These are not great terms, but something along those lines might be clearer.

Reply:
I think a good word to use is "accessible" and "inaccessible", at which point this comment becomes:
```rs
/// Wraps a mutably borrowed value of type `T`.
///
/// This prevents all other borrows of `value` while this guard is accessible. To make this guard
/// inaccessible, use [`GdCell::make_inaccessible()`](crate::GdCell::make_inaccessible) on a mutable
/// reference handed out by this guard.
```

Since the way the non-aliasing guard works is that it makes the value inaccessible while the guard is alive. And the fact that the value is inaccessible is what lets us reborrow it. 

I can also reword things to talk about what must be true while a value is accessible vs inaccessible.

Accessible and inaccessible are also more familiar and easy to understand words than aliasing vs non-aliasing.

Reply:
> And the fact that the value is inaccessible is what lets us reborrow it.

I'm not sure this wording is really better; this in particular seems counter-intuitive. _Reading_ the value may be inaccessible, but reborrowing itself is a form of access.

Maybe we should not try to come up with any new terms, because they inevitably require that two terms (reborrowing + X) are associated, needing the user to understand this concept first.

Why not directly call it what it is/does: `set_reborrowed`?

Reply:
The guard is what's made inaccessible. The value is still accessible through the cell. You make the guard inaccessible so that you can borrow from the cell again.

Also it would be something more like `set_reborrowable`, since it doesn't get reborrowed immediately upon the call to this method. It just allows you to borrow from the cell again without dropping the guard.